registry/decoder.go

```diff
@@ -0,0 +1,79 @@
+package registry
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+
+	"github.com/hamba/avro/v2"
+)
+
+// Decoder is entitled to receive raw messages as bytes, obtain the
+// related schema and then use it to deserialize the remaining content.
+type Decoder struct {
```

Comment:
The ordering here is a little weird. Conventionally, it is best where possible to keep the struct and its constructor together. So I would order like so (in short form):
```
type DecoderFunc

func WithAPI

type Decoder

func NewDecoder

// all the other code ...

func extractSchemaIDFromPayload
```

The unattached function can either be at the end or just above `Decode`, dealers choice.